lib/src/Uncertainty/Process/DiscreteMarkovChain.cxx

```diff
@@ -0,0 +1,243 @@
+#include "openturns/Point.hxx"
+#include "openturns/SpecFunc.hxx"
+
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(DiscreteMarkovChain)
+
+static const Factory<DiscreteMarkovChain> Factory_DiscreteMarkovChain;
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain()
+  : ProcessImplementation()
+  , origin_(0)
+  , transitionMatrix_(SquareMatrix(1, 1))
+  , currentState_(0)
+{
+  setOutputDimension(1);
+}
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain(const UnsignedInteger origin,
+                       const SquareMatrix & transitionMatrix)
+  : ProcessImplementation()
+  , origin_(origin)
+  , transitionMatrix_(transitionMatrix)
+  , currentState_(origin)
+{
+  // Set the dimension of the process
+  setOutputDimension(1);
+}
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain(const UnsignedInteger origin,
+                       const SquareMatrix & transitionMatrix,
+                       const RegularGrid & timeGrid)
+  : ProcessImplementation()
+  , origin_(origin)
+  , transitionMatrix_(transitionMatrix)
+  , currentState_(origin)
+{
+  setOutputDimension(transitionMatrix.getNbRows());
+  setTimeGrid(timeGrid);
+}
+
+/* Virtual constructor */
+DiscreteMarkovChain * DiscreteMarkovChain::clone() const
+{
+  return new DiscreteMarkovChain(*this);
+}
+
+/* String converter */
+String DiscreteMarkovChain::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << DiscreteMarkovChain::GetClassName()
+      << ", origin=" << origin_
+      << ", transition matrix=" << transitionMatrix_;
+  return oss;
+}
+
+/* String converter */
+String DiscreteMarkovChain::__str__() const
+{
+  OSS oss;
+  oss << getClassName() << "(origin=" << origin_ << ", transition=" << PersistentCollection<Scalar>(*transitionMatrix_.getImplementation()).__str__() << ")";
+  return oss;
+}
+
+/* Realization accessor */
+Field DiscreteMarkovChain::getRealization() const
+{
+  const UnsignedInteger size = mesh_.getVerticesNumber();
+  const UnsignedInteger dimension = transitionMatrix_.getNbRows();
+
+  
+  /* Definition of the states of the system */
+  Sample states(dimension, 1);
+  for (UnsignedInteger i = 0; i < dimension; i++)
+  {
+    states(i, 0) = i;
+  }
+
+  /* Loop on mesh points */
+  Sample data(size, 1);
+  data(0, 0) = origin_;
+  UserDefined dist(states);
+
+  for (UnsignedInteger i = 1; i < size; i++)
+  {
+    /* Definition of the distribution */
+    Point weights(dimension, 0.0);
+    for (UnsignedInteger col = 0; col < dimension; col++)
+    {
+      weights[col] = transitionMatrix_(currentState_, col);
+    }
+    /* Get a realization of the distribution */
+    dist.setData(states, weights);
+    data[i] = dist.getRealization();
+    currentState_ = data(i, 0);
+  }
+  return Field(mesh_, data);
+}
+
+/* Compute the next steps of a Markov chain */
+TimeSeries DiscreteMarkovChain::getFuture(const UnsignedInteger stepNumber) const
+{
+  // TimeGrid of the process
+  RegularGrid timeGrid;
+  try
```

Comment:
This check should be done once and for all in the accessor to the underlying mesh. As a rule, a DiscreteRandomWalk is defined on a RegularGrid, so setMesh() should be overloaded to check this property.